runners/reference/java/src/main/java/org/apache/beam/runners/reference/JobServicePipelineResult.java

```diff
@@ -121,12 +140,110 @@ public State waitUntilFinish() {
 
   @Override
   public MetricResults metrics() {
-    throw new UnsupportedOperationException("Not yet implemented.");
+    if (metricResults == null) {
```

Comment:
The semantics are currently that this should be called once the state is `DONE`, and at that time it fetches + caches the `MetricsResults` object for this job from the job server.

TODO: bypass caching the `MetricResults` object until the state becomes `DONE`, so that `metrics()` can be polled while a job is running.